src/main/java/org/chocosolver/solver/constraints/set/PropIntersectionFilterSets.java

```diff
@@ -0,0 +1,122 @@
+        }
+        if (sets.length == 1) {
+            throw new IllegalArgumentException("This propagator does not work for a single set, "
+                    + "use new PropAllEqual(new SetVar[]{sets[0], intersection}) instead.");
+        }
+        k = sets.length;
+        sdm = new ISetDeltaMonitor[k + 1];
+        for (int i = 0; i <= k; i++) {
+            sdm[i] = this.vars[i].monitorDelta(this);
+        }
+    }
+
+    @Override
+    public int getPropagationConditions(int vIdx) {
+        if (vIdx < k) {
+            return SetEventType.ADD_TO_KER.getMask();
+        } else {
+            return SetEventType.REMOVE_FROM_ENVELOPE.getMask();
+        }
+    }
+
+    private SetVar findSetThatDoesNotContainJInLB(int j) {
+        for (int i = 0; i < k - 1; i++) {
+            if (!vars[i].getLB().contains(j)) {
+                return vars[i];
+            }
+        }
+        return vars[k - 1];
+    }
+
+    @Override
+    public void propagate(int evtmask) throws ContradictionException {
+        SetVar intersection = vars[k];
+        // Maps elements in sets to the number of occurences of those elements.
+        TIntIntHashMap count = new TIntIntHashMap();
+        for (int i = 0; i < k; i++) {
+            ISetIterator iter = vars[i].getLB().iterator();
+            while (iter.hasNext()) {
+                int j = iter.nextInt();
+                if (!intersection.getUB().contains(j)) {
+                    if (count.adjustOrPutValue(j, 1, 1) >= k - 1) {
+                        // intersection does not contain j but k - 1 sets
+                        // contains j, hence the last set must not contain j.
+                        findSetThatDoesNotContainJInLB(j).remove(j, this);
+                    }
+                }
+            }
+        }
+        for (int i = 0; i <= k; i++) {
+            sdm[i].unfreeze();
+        }
+    }
+
+    private SetVar findUniqueSetThatDoesNotContainJInLB(int j) {
+        SetVar uniqueSet = null;
+        for (int i = 0; i < k; i++) {
+            if (!vars[i].getLB().contains(j)) {
+                if (uniqueSet != null) {
+                    return null;
+                }
+                uniqueSet = vars[i];
+            }
+        }
+        return uniqueSet;
+    }
+
+    @Override
+    public void propagate(int idxVarInProp, int mask) throws ContradictionException {
+        sdm[idxVarInProp].freeze();
+        if (idxVarInProp < k) {
+            SetVar intersection = vars[k];
+            sdm[idxVarInProp].forEach(j -> {
+                if (!intersection.getUB().contains(j)) {
+                    SetVar uniqueSet = findUniqueSetThatDoesNotContainJInLB(j);
+                    if (uniqueSet != null) {
+                        uniqueSet.remove(j, this);
+                    }
+                }
+            }, SetEventType.ADD_TO_KER);
+        } else {
+            sdm[idxVarInProp].forEach(j -> {
```

Comment:
Does this create a new lambda at each propagation? Doesn't it have an overhead compared to creating an object once and for all in the constructor?